api/internal/git/repospec.go

```diff
@@ -209,7 +215,15 @@ func peelQuery(arg string) (string, string, time.Duration, bool) {
 		}
 	}
 
-	return parsed.Path, ref, duration, submodules
+	return ref, duration, submodules
+}
+
+func parsePath(n string) string {
+	parsed, err := url.Parse(n)
+	if err != nil {
+		return ""
```

Comment:
I'm having a bit of trouble following this. If the url is not valid, why do we return an empty string instead of erroring out? A comment here might be helpful

Reply:
Comment added. The previous behavior of `peelQuery`, which also parsed the path, was to return defaults in the case of any error.

I actually wanted to discuss this. It also seems weird to me that we silently correct for error cases. Failing fast might help users realize their mistakes earlier and would help distinguish between an intention url and a local path resembling a url. However, we do have test cases covering this silent failure behavior.

Reply:
> However, we do have test cases covering this silent failure behavior.

Which do you mean? If I change this to `return "oops"` I don't see anything failing currently (in the repo spec suite at least). The simplest way I found to trigger it is path ending in a percent sign `http://github.com/someorg/somerepo/%`. 

The more I think about this, the more I'm inclined to favour either the `return ""` you have here or an actual error (or maybe, log a warning but continue with the empty return?). Since the value here ultimately gets used to make a request, preserving things that are invalid in an unknown way feels like a potential vulnerability vector.

Reply:
Good catch! Yes, I jumped to the wrong conclusion reading this comment: https://github.com/kubernetes-sigs/kustomize/pull/4863/files#diff-109bb418d3031aaee374127b42b05a6845a4fb528a637c734a84270be49d8c8fR182

I've corrected this and added a corresponding test.

Reply:
> > However, we do have test cases covering this silent failure behavior.
> 
> Which do you mean?

I was originally referring to the "silent failure behavior" of `Ref`, `Submodules`, and `Timeout`. If we [encounter an error parsing the query](https://github.com/kubernetes-sigs/kustomize/pull/4863/files#diff-109bb418d3031aaee374127b42b05a6845a4fb528a637c734a84270be49d8c8fL177) (url.Parse() includes code parsing the query) or find unexpected query parameters or values, we return the defaults for these fields, which I found surprising.

Reply:
> The more I think about this, the more I'm inclined to favour either the `return ""` you have here or an actual error (or maybe, log a warning but continue with the empty return?). Since the value here ultimately gets used to make a request, preserving things that are invalid in an unknown way feels like a potential vulnerability vector.

Yeah, these options generally sound good to me! I'm hesitant to include the entirety of the error-handling fix here because I see several cases we need to handle

* `url.ParseQuery()` error, in which case I'm inclined to throw an error
* `url.Parse()` error when the output is intended for `Path`, in which case the error is more forgivable because there are many illegal uri characters allowed in file paths
* `url.Parse()` error when the output is intended for `OrgRepo`, a case I'd have to further investigate because I don't believe `git fetch` urls follow [rfc3986 uri syntax](https://www.rfc-editor.org/rfc/rfc3986) to a tee

WDYT about throwing the error in `parseQuery` and leaving a TODO statement before the error-handling in `parsePath` for a future PR? I set out in this PR to properly extract the query and preserve all other behavior (changed the PR title).

Reply:
That works for me. I agree that it makes sense to focus on the error handling change on its own, if we want to make it. Maybe add it to the list on your issue as well?

Reply:
Done